botocore/utils.py

```diff
@@ -644,24 +646,32 @@ def fix_s3_host(request, signature_version, region_name, **kwargs):
         return
     elif not _allowed_region(region_name):
         return
+
+    # We need to respect custom endpoint urls, but unfortunately we have no
+    # way at this point to know how the url got set. We unfortunately don't have
+    # access to the endpoint resolver here, so we can't do a data-driven test.
+    # A regular expression is therefore used to determine if the request url is
+    # an S3 url.
+    endpoint_url = None
+    match = S3_ENDPOINT_RE.match(request.url)
+    if match is not None:
+        endpoint_url = 's3.amazonaws.com'
+
     try:
-        switch_to_virtual_host_style(
-            request, signature_version, 's3.amazonaws.com')
+        switch_to_virtual_host_style(request, endpoint_url)
     except InvalidDNSNameError as e:
         bucket_name = e.kwargs['bucket_name']
         logger.debug('Not changing URI, bucket is not DNS compatible: %s',
                      bucket_name)
 
 
-def switch_to_virtual_host_style(request, signature_version,
```

Comment:
This is arguably a breaking change by removing the variable. I think that I am fine with it because I do not see users trying to use this method since we already have configuration options. @jamesls what are your thoughts.